plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/server/debug/OpenShiftDebugMode.java

```diff
@@ -0,0 +1,414 @@
+			Collection<IContainer> containers = new ArrayList<>(originalContainers);
+			IContainer container = containers.iterator().next();
+			Set<IPort> ports = new HashSet<>(container.getPorts());
+			
+			IPort existing = ports.stream()
+					.filter(p -> p.getContainerPort() == debugPort)
+					.findFirst()
+					.orElse(null);
+			boolean added = addDebugPort(debugPort, ports, existing);
+			if (added) {
+				container.setPorts(ports); 
+				dc.setContainers(containers);
+			}
+		}
+	}
+
+	private static boolean addDebugPort(int debugPort, Set<IPort> ports, IPort existing) {
+		boolean added = false;
+		if (existing == null) {
+			PortSpecAdapter newPort = new PortSpecAdapter("debug", "TCP", debugPort);
+			added = ports.add(newPort);
+		} else {
+			PortSpecAdapter newPort = new PortSpecAdapter(existing.getName(), existing.getProtocol(), debugPort);
+			if (!existing.equals(newPort)) {
+				ports.remove(existing);
+				added = ports.add(newPort);
+			}
+		}
+		return added;
+	}
+
+	private static boolean isDebugEnabled(IDeploymentConfig dc, String devmodeKey, String debugPortKey) {
+		boolean debugEnabled = false;
+		boolean devmodeEnabled = isDevmodeEnabled(dc, devmodeKey);
+		if (devmodeEnabled) {
+			// debugging is enabled if devmode and debug port are set
+			String debugPort = getEnv(dc, debugPortKey);
+			debugEnabled = !StringUtils.isBlank(debugPort);
+		}
+		return debugEnabled;
+	}
+
+	/**
+	 * Enables devmode in given context.  No change is
+	 * executed in OpenShift, the change is only in the given local context. To have
+	 * the changes set to OpenShift one has to call
+	 * {@link #sendChanges(DebugContext, IProgressMonitor)}
+	 * 
+	 * @param context
+	 * @param monitor
+	 * @throws CoreException
+	 */
+	public static void enableDevmode(DebugContext context) {
+		context.setDevmodeEnabled(true);
+	}
+	
+	/**
+	 * Disables devmode in given context.  No change is
+	 * executed in OpenShift, the change is only in the given local context. To have
+	 * the changes set to OpenShift one has to call
+	 * {@link #sendChanges(DebugContext, IProgressMonitor)}
+	 * 
+	 * @param context
+	 * @param monitor
+	 * @throws CoreException
+	 */
+	public static void disableDevmode(DebugContext context) {
+		context.setDevmodeEnabled(false);
+	}
+
+	/**
+	 * Sends the changes to the deployment config if required. Nothing is done if
+	 * the deployment config already is in the state that the context is in. If the
+	 * changes are sent the existing pods are killed and it waits until the new pods
+	 * are running.
+	 * 
+	 * @param context
+	 * @param monitor
+	 * @throws CoreException
+	 */
+	public static void sendChanges(DebugContext context, IProgressMonitor monitor) throws CoreException {
+		IDeploymentConfig dc = getDeploymentConfig(context, monitor);
+		if (dc == null) {
+			throw new CoreException(StatusFactory.errorStatus(
+					OpenShiftCoreActivator.PLUGIN_ID, "No deployment config present that can be updated."));
+		}
```

Comment:
this complaint is false! the "|" is used on purpose so that both bethods are ALWAYS executed.